src/backend/metal/src/command.rs

```diff
@@ -2160,117 +2172,50 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
-                            let res = &layout.res_overrides[&msl::ResourceBindingLocation {
-                                binding: binding as _,
-                                .. location_fs
-                            }];
-                            match *values {
-                                Sampler(ref samplers) => {
-                                    let start = res.sampler_id as usize;
-                                    resources.add_samplers(start, samplers.as_slice());
-                                    commands.extend(samplers.iter().cloned().enumerate().map(|(i, sampler)| {
-                                        soft::RenderCommand::BindSampler {
-                                            stage: pso::Stage::Fragment,
-                                            index: start + i,
-                                            sampler,
+                            Buffer(ref buffers) => {
+                                for (i, bref) in buffers.iter().enumerate() {
+                                    let (buffer, offset) = match bref.base {
+                                        Some((ref buffer, mut offset)) => {
+                                            if bref.dynamic {
+                                                offset += *offset_iter
+                                                    .next()
+                                                    .expect("No dynamic offset provided!")
+                                                    .borrow() as u64;
+                                            }
+                                            (Some(buffer), offset)
                                         }
-                                    }));
-                                }
-                                Image(ref images) => {
-                                    let start = res.texture_id as usize;
-                                    resources.add_textures(start, images.as_slice());
-                                    commands.extend(images.iter().enumerate().map(|(i, texture)| {
-                                        soft::RenderCommand::BindTexture {
-                                            stage: pso::Stage::Fragment,
-                                            index: start + i,
-                                            texture: texture.as_ref().map(|&(ref texture, _)| texture.clone()),
+                                        None => (None, 0),
+                                    };
+                                    for &mut (ref loc, ref mut resources) in &mut bind_stages {
+                                        let start = layout.res_overrides[loc].buffer_id as usize;
+                                        if let Some(buffer) = buffer {
+                                            resources.add_buffer(start + i, buffer, offset as _);
                                         }
-                                    }));
```

Comment:
What happened to all the previous `commands.extend`/`push` here? Are they not actually required?

Reply:
The old code was doing exactly the same thing for vertex and fragment stages. The problem came from a case where a dynamic buffer is used in both, so the code would try to advance the iterator twice for the same buffer...

The new code just builds a `bind_stages` vector (of size 2 at max) beforehand, and iterates it for each descriptor set type, thus bringing the amount of code duplication to a reasonable minimum as well as resolving the dynamic buffer issue.